### ISSUE TITLE
Include fake media query service for easy testing with bUnit in package

### ISSUE THREAD

**Issue by egil** [EdCharbeneau/BlazorSize#50]
After looking at #49, I have the following suggestion.

Include something like the following two classes in the BlazorSize assembly:

```c#
using System.Collections.Generic;
using System.Threading.Tasks;
using BlazorPro.BlazorSize;
using Microsoft.JSInterop;

namespace Bunit
{
    public class FakeMediaQueryService : IMediaQueryService
    {
        private readonly TestContextBase contextBase;
        private MediaQueryArgs? activeMediaQuery = null;

        public List<MediaQueryCache> MediaQueries { get; } = new();

        public FakeMediaQueryService(TestContextBase contextBase)
        {
            this.contextBase = contextBase;
        }

        public void AddQuery(MediaQuery newMediaQuery)
        {
            var mediaQueryCache = GetMediaQueryFromCache(newMediaQuery.Media);

            if (mediaQueryCache is null)
            {
                mediaQueryCache = new MediaQueryCache()
                {
                    MediaRequested = newMediaQuery.Media
                };
                MediaQueries.Add(mediaQueryCache);
            }

            mediaQueryCache.MediaQueries.Add(newMediaQuery);

            SetActiveMediaQuery(activeMediaQuery);
        }

        public Task CreateMediaQueryList(DotNetObjectReference<MediaQueryList> dotNetObjectReference)
            => Task.CompletedTask;

        public Task Initialize(MediaQuery mediaQuery)
            => Task.CompletedTask;

        public Task RemoveQuery(MediaQuery mediaQuery)
        {
            var mediaQueryFromCache = GetMediaQueryFromCache(mediaQuery.Media);
            if (mediaQueryFromCache is null)
                return Task.CompletedTask;

            mediaQueryFromCache.MediaQueries.Remove(mediaQuery);

            if (mediaQueryFromCache.MediaQueries.Count == 0)
                MediaQueries.Remove(mediaQueryFromCache);

            return Task.CompletedTask;
        }

        public void SetActiveBreakPoint(string breakpoint)
            => SetActiveMediaQuery(new MediaQueryArgs
            {
                Matches = true,
                Media = breakpoint
            });

        public void SetActiveMediaQuery(MediaQueryArgs args)
        {
            if (args is null)
                return;

            activeMediaQuery = args;

            // cache must be compared by actual value, not RequestedMedia when invoked from JavaScript
            // DOM Media value my be different that the initally requested media query value.
            var cache = GetMediaQueryFromCache(args.Media);

            if (cache is null) return;

            // Dispatch events to all subscribers
            contextBase.Renderer.Dispatcher.InvokeAsync(() =>
            {
                foreach (var item in cache.MediaQueries)
                {
                    item.MediaQueryChanged(args);
                }
            });
        }

        private MediaQueryCache GetMediaQueryFromCache(string byMedia)
            => MediaQueries?.Find(q => q.MediaRequested == byMedia);
    }
}
```

and 

```c#
using BlazorPro.BlazorSize;
using Microsoft.Extensions.DependencyInjection;

namespace Bunit
{
    public static class BunitExtensions
    {
        public static FakeMediaQueryService AddBlazorSize(this TestContextBase testContext)
        {
            testContext.RenderTree.Add<MediaQueryList>();
            var mediaQueryServices = new FakeMediaQueryService(testContext);
            testContext.Services.AddSingleton<IMediaQueryService>(mediaQueryServices);
            return mediaQueryServices;
        }
    }
}
```

I say "something like", since I have not tested all scenarios, and might not have correct understanding of how the library works. 

This will enable us to write a test of the following `<MyComp>` component:

```razor
@using BlazorPro.BlazorSize
<MediaQuery Media="@Breakpoints.SmallDown">
    <Matched>
        MATCHES
    </Matched>
    <Unmatched>
        DOES NOT MATCH
    </Unmatched>
</MediaQuery>
```

The test looks like this:

```c#
using BlazorPro.BlazorSize;
using Bunit;
using Xunit;

namespace BlazorSizeAndBunit
{
    public class MyCompTest
    {
        [Fact]
        public void Shows_Correct_Content_Based_On_MediaQuery()
        {
            // Arrange
            using var ctx = new TestContext();
            var blazorSizeState = ctx.AddBlazorSize();
            blazorSizeState.SetActiveBreakPoint(Breakpoints.LargeDown);

            // Act
            var cut = ctx.RenderComponent<MyComp>();

            // Assert
            cut.MarkupMatches("DOES NOT MATCH");

            // Change breakpoint
            blazorSizeState.SetActiveBreakPoint(Breakpoints.SmallDown);

            // Assert
            cut.MarkupMatches("MATCHES");
        }
    }
}
```

This is what's going on here:

1. The extension method above enables us to easily add support for BlazorSize to bUnit.
2. The method `SetActiveBreakPoint` allows us to specify an active breakpoint to report to the BlazorSize components.
3. When ever we call `SetActiveBreakPoint`, the BlazorSize componentsare notified and the component under test re-renders to show new content.

Hope this makes sense, otherwise, ask away.

**Comment by egil** [comment]
If there is no need for (test) users to inspect or call all the methods on `IMediaQueryService`, you could choose to not return the `FakeMediaQueryService` from `AddBlazorSize` and instead return a `FakeBrowserWindow` or similarly named type, that has properties such as `Width` or `Breakpoint`, and users would simply specify a width or breakpoint on that, and it would pass the correct `MediaQueryArgs` to the `FakeMediaQueryService`.

**Comment by egil** [comment]
Looks good to me. Only thing I would probably change is to reference bunit.core instead of bunit.web in https://github.com/EdCharbeneau/BlazorSize/blob/master/BlazorPro.BlazorSize.Bunit/BlazorPro.BlazorSize.Bunit.csproj#L8. But its a minor thing tbh.

**Comment by EdCharbeneau** [comment]
@egil Thanks for the review, you're too kind :)

**Comment by EdCharbeneau** [comment]
Fixed in 5.0 https://github.com/EdCharbeneau/BlazorSize/releases/tag/5.0

**Comment by egil** [comment]
> Fixed in 5.0 https://github.com/EdCharbeneau/BlazorSize/releases/tag/5.0

Simple and easy.
